bin/typed_code.py

```diff
@@ -0,0 +1,303 @@
+            teams.update(TEAM_REGEX.findall(line))
+
+    logger.debug("All teams")
+    logger.debug("\n".join(teams))
+    return teams
+
+
+def split_files_by_codeowner(files: Set[str], codeowners: Any) -> MutableMapping[str, Set[str]]:
+    """
+    Given a list of filenames and a codeowners objects, split the files up by
+    owner. This isn't a full solution because it doesn't handle multiple owners
+    on a file.
+    """
+    files_by_codeowner = defaultdict(set)
+    for filename in files:
+        owners = codeowners.of(filename)
+        logger.debug(f"{filename} {owners}")
+
+        owners = {owner[1] for owner in owners} if owners else {UNOWNED_KEY}
+        for owner in owners:
+            files_by_codeowner[owner].add(filename)
+    return files_by_codeowner
+
+
+def load_cache(filename: Optional[str] = None) -> MutableMapping[str, int]:
+    logger.debug(f"loading cache from {filename}")
+
+    if not (filename and os.path.exists(filename)):
+        logger.debug("file not found")
+        return {}
+
+    cache = {}
+    with open(filename) as f:
+        try:
+            for line in f.readlines():
+                key, value = line.split(CACHE_SEPARATOR)
+                cache[key] = int(value)
+        except (AttributeError, OSError, TypeError, ValueError):
+            return {}
+    return cache
+
+
+def store_cache(cache: Mapping[str, int], filename: str) -> None:
+    # TODO We don't garbage collect stale hashes so the file cache will continue
+    #  to grow indefinitely.
+    if not filename:
+        return
+
+    with open(filename, "w") as f:
+        for key, value in cache.items():
+            f.write(f"{key}{CACHE_SEPARATOR}{value}\n")
+
+
+def hash_file(filename: str) -> str:
+    """https://stackoverflow.com/questions/22733826"""
+    func = hashlib.md5()
+    with open(filename, "rb") as f:
+        while True:
+            block = f.read(1024 * func.block_size)
+            if not block:
+                break
+            func.update(block)
+    return func.hexdigest()
+
+
+def analyze_file(file: str, cache: MutableMapping[str, int]) -> int:
+    """Evan"s algorithm for grabbing LOC from a file."""
+    filename = os.path.join(BASE_DIR, file)
+
+    key = hash_file(filename)
+    cached_value = cache.get(key)
+    if cached_value is not None:
+        logger.debug(f"cache hit {filename}")
+        return cached_value
+
+    logger.debug(f"cache size {len(cache.keys())}")
+    logger.debug(f"cache miss {filename} {key}")
+    proc_cmd = f"pygount {filename} --format=summary --suffix=py"
+    proc = subprocess.run(proc_cmd.split(" "), capture_output=True)
+    output = proc.stdout.decode("utf-8")
+    value = int(output.split("\n")[-2].split()[-2])
+
+    cache[key] = value
+    return value
+
+
+def total_lines(files: Set[str], cache: MutableMapping[str, int], status: str = "") -> int:
+    """Gets the total lines and primes the cache."""
+    total = 0
+    for i, file in enumerate(files):
```

Comment:
Probably this is so slow because you are running a subprocess for every single file in the repo when you analyze the entire ROOT directory. In my original script I ran pygount for the entire repo in a single command, and then just on the directories instead of individual files.

Reply:
The `pygount` report has LOC per file! We can run it once and cache the results.